File: src/lib/ciphers/juniper9.ts
/**
 * Juniper $9$ reversible "encryption".
 *
 * $9$ is a keyless, device-independent substitution cipher: a value encoded on
 * one Juniper device decodes on any other with no node-specific secret. It is
 * obfuscation, not cryptography - anyone with this code can recover the
 * plaintext. Ported from the public Crypt::Juniper / junosdecode algorithm.
 */

import type { Cipher } from "./types";

const MAGIC = "$9$";
const FAMILY = [
  "QzF3n6/9CAtpu0O",
  "B1IREhcSyrleKvMW8LXx",
  "7N-dVbwsY2g4oaJZGUDj",
  "iHkq.mPf5T",
];
const ALPHA = FAMILY.join("");
const SIZE = ALPHA.length;

const NUM: Record<string, number> = {};
for (let i = 0; i < ALPHA.length; i++) {
  NUM[ALPHA[i]] = i;
}

const EXTRA: Record<string, number> = {};
FAMILY.forEach((family, i) => {
  for (const c of family) {
    EXTRA[c] = 3 - i;
  }
});

const ENCODING = [
  [1, 4, 32],
  [1, 16, 32],
  [1, 8, 32],
  [1, 64],
  [1, 32],
  [1, 4, 16, 128],
  [1, 32, 64],
];

/** Positive modulo (JS `%` keeps the sign of the dividend). */
function mod(a: number, n: number): number {
  return ((a % n) + n) % n;
}

function gap(c1: string, c2: string): number {
  return mod(NUM[c2] - NUM[c1], SIZE) - 1;
}

export function decode(encoded: string): string {
  encoded = encoded.trim();
  if (!encoded.startsWith(MAGIC)) {
    throw new Error("Not a Juniper $9$ string: must start with $9$");
  }
  let chars = encoded.slice(MAGIC.length);
  if (chars.length === 0) {
    throw new Error("Empty $9$ payload");
  }
  const first = chars[0];
  if (!(first in NUM)) {
    throw new Error(`Invalid start character '${first}': not in $9$ alphabet`);
  }
  chars = chars.slice(1 + EXTRA[first]);
  let prev = first;
  const result: string[] = [];
  while (chars.length > 0) {
    const weights = ENCODING[result.length % ENCODING.length];
    const group = chars.slice(0, weights.length);
    if (group.length < weights.length) {
      throw new Error("Truncated $9$ ciphertext: incomplete final character group");
    }
    chars = chars.slice(weights.length);
    let val = 0;
    for (let k = 0; k < weights.length; k++) {
      const c = group[k];
      if (!(c in NUM)) {
        throw new Error(`Character '${c}' not in $9$ alphabet: ciphertext may be invalid`);
      }
      val += gap(prev, c) * weights[k];
      prev = c;
    }
    result.push(String.fromCharCode(mod(val, 256)));
  }
  return result.join("");
}

export function encode(plaintext: string): string {
  const randChar = () => ALPHA[Math.floor(Math.random() * SIZE)];
  const start = randChar();
  const result: string[] = [MAGIC, start];
  for (let i = 0; i < EXTRA[start]; i++) {
    result.push(randChar());
  }
  let prev = start;
  for (let i = 0; i < plaintext.length; i++) {
    const weights = ENCODING[i % ENCODING.length];
    let val = plaintext.charCodeAt(i);
    if (val > 255) {
      throw new Error(
        `Character '${plaintext[i]}' cannot be $9$-encoded: only single-byte (Latin-1) characters are supported`,
      );
    }
    const gaps: number[] = [];
    for (let j = weights.length - 1; j >= 0; j--) {
      gaps.push(Math.floor(val / weights[j]));
      val = val % weights[j];
    }
    gaps.reverse();
    for (const g of gaps) {
      const c = ALPHA[mod(NUM[prev] + g + 1, SIZE)];
      result.push(c);
      prev = c;
    }
  }
  return result.join("");
}

export const juniper9: Cipher = {
  id: "juniper9",
  index: "01",
  name: "Juniper $9$",
  vendor: "Juniper",
  magic: "$9$",
  tagline: "Juniper type 9 ($9$): reversible, keyless obfuscation for JUNOS secrets.",
  reversible: true,
  keyed: false,
  status: "available",
  notes: [
    "$9$ is a substitution cipher, not real cryptography: it is reversible by anyone, with no key.",
    "It is device-independent - a value encoded on one Juniper box decodes on any other.",
    "Encoding is non-deterministic: random filler means the same plaintext yields a different $9$ string each time.",
    "Treat it as obfuscation against shoulder-surfing, never as protection.",
  ],
  example: {
    encoded: "$9$FNkC3/t1IcevLuOWx",
    plaintext: "hello",
  },
  links: [
    {
      label: "github.com/antoinekh/juniper9-crypt",
      url: "https://github.com/antoinekh/juniper9-crypt",
      note: "Python library & CLI, plus the documented algorithm.",
    },
  ],
  encode: async (plaintext) => encode(plaintext),
  decode: async (encoded) => decode(encoded),
};

File: network_secret/juniper9.py
"""
Encrypt and decrypt Juniper/HPE $9$ reversible passwords.

The $9$ algorithm is a proprietary Juniper/HPE substitution cipher that is
deterministic and device-independent: the same plaintext can be encrypted
on any Juniper/HPE device and decrypted on any other, with no node-specific
secret involved.

This is a Python implementation of the algorithm. It is based on
`Crypt::Juniper` (Perl, Kevin Brintnall) and Matt Hite's Python 2 port
at https://github.com/mhite/junosdecode.

Public API:
    decrypt(ciphertext: str) -> str
    encrypt(plaintext: str) -> str
    check(ciphertext: str, other: str) -> tuple[str, str, bool]
"""

from __future__ import annotations

import random

__all__ = ["decrypt", "encrypt", "check"]

MAGIC = "$9$"
FAMILY = ["QzF3n6/9CAtpu0O", "B1IREhcSyrleKvMW8LXx", "7N-dVbwsY2g4oaJZGUDj", "iHkq.mPf5T"]
EXTRA = {c: 3 - i for i, f in enumerate(FAMILY) for c in f}
ALPHA = "".join(FAMILY)
NUM = {c: i for i, c in enumerate(ALPHA)}
ENCODING = [[1, 4, 32], [1, 16, 32], [1, 8, 32], [1, 64], [1, 32], [1, 4, 16, 128], [1, 32, 64]]


def _gap(c1: str, c2: str) -> int:
    return (NUM[c2] - NUM[c1]) % len(ALPHA) - 1


def decrypt(ciphertext: str) -> str:
    """Decrypt a Juniper/HPE $9$ ciphertext back to plaintext."""
    if not ciphertext.startswith(MAGIC):
        raise ValueError(f"Not a Juniper/HPE $9$ string: must start with {MAGIC}")
    chars = ciphertext[len(MAGIC):]
    if not chars:
        raise ValueError("Empty $9$ payload")
    first = chars[0]
    if first not in NUM:
        raise ValueError(f"Invalid start character {first!r}: not in $9$ alphabet")
    chars = chars[1 + EXTRA[first]:]
    prev = first
    result: list[str] = []
    while chars:
        weights = ENCODING[len(result) % len(ENCODING)]
        nibble = chars[: len(weights)]
        if len(nibble) < len(weights):
            raise ValueError("Truncated $9$ ciphertext: incomplete final character group")
        chars = chars[len(weights):]
        val = 0
        for c, w in zip(nibble, weights):
            if c not in NUM:
                raise ValueError(f"Character {c!r} not in $9$ alphabet: ciphertext may be invalid")
            val += _gap(prev, c) * w
            prev = c
        result.append(chr(val % 256))
    return "".join(result)


def encrypt(plaintext: str) -> str:
    """Encrypt plaintext to a Juniper/HPE $9$ ciphertext.

    Output is non-deterministic: random filler characters mean each call
    returns a different ciphertext for the same plaintext. All of them
    decrypt back to the same plaintext. `random` (not `secrets`) is used
    deliberately, since $9$ is a substitution cipher with no real
    cryptographic strength: a stronger RNG buys nothing.
    """
    start = random.choice(ALPHA)
    result = [MAGIC, start]
    for _ in range(EXTRA[start]):
        result.append(random.choice(ALPHA))
    prev = start
    for i, ch in enumerate(plaintext):
        weights = ENCODING[i % len(ENCODING)]
        val = ord(ch)
        if val > 255:
            raise ValueError(
                f"Character {ch!r} cannot be $9$-encoded: only single-byte (Latin-1) characters are supported"
            )
        gaps: list[int] = []
        for w in reversed(weights):
            gaps.append(val // w)
            val = val % w
        for gap in reversed(gaps):
            c = ALPHA[(NUM[prev] + gap + 1) % len(ALPHA)]
            result.append(c)
            prev = c
    return "".join(result)


def check(ciphertext: str, other: str) -> tuple[str, str, bool]:
    """Decrypt `ciphertext` and compare to `other`.

    `other` may be a plaintext, or another $9$ value (auto-detected by
    the $9$ prefix). Returns (plain_a, plain_b, match).
    """
    plain_a = decrypt(ciphertext)
    plain_b = decrypt(other) if other.startswith(MAGIC) else other
    return plain_a, plain_b, plain_a == plain_b
